### PR TITLE
wolfsshd: refuse sessions it cannot serve

### PR DESCRIPTION
**Pull request by ejohnstown**
A shell, exec or subsystem request is answered as it arrives, so a session this build cannot serve is refused with CHANNEL_FAILURE rather than accepted and then dropped once the session is up.

- `SessionRequestCb()` takes a shell with `WOLFSSH_SHELL`, an exec with `WOLFSSH_SHELL` or an scp command with `WOLFSSH_SCP`, and the sftp subsystem with `WOLFSSH_SFTP`; anything else is refused and logged.
- A request whose command did not fit is refused rather than read through a `NULL`.
- `wolfSSH_ChannelCommandIsScp()` takes "scp" only as its own token and is shared with the SCP divert in `wolfSSH_accept()`, so an exec of `scpbackup foo` no longer reaches the built-in SCP server.
- `wolfSSH_ChannelGetSessionGranted()` reports a channel's grant, so an application need not read `WOLFSSH_CHANNEL`.
- sshd_bad_subsystem_test.sh asks for an unknown subsystem with the OpenSSH client and expects the refusal.

What the daemon serves is otherwise unchanged.

### CODE REVIEW

File: apps/wolfsshd/wolfsshd.c [hunk 1]
```diff
@@ -359,6 +359,80 @@ static void CleanupCTX(WOLFSSHD_CONFIG* conf, WOLFSSH_CTX** ctx,
...
+        #endif
+            break;
+
+        case WOLFSSH_SESSION_EXEC:
+            if (cmd == NULL) {
+                reason = "exec request carried no command";
+                break;
+            }
+        #ifdef WOLFSSH_SCP
+            if (WSTRNCMP(cmd, "scp", 3) == 0) {
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply apps/wolfsshd/wolfsshd.c]: The prefix match is deliberate: it mirrors `ChannelCommandIsScp()` in `src/wolfscp.c`, which uses the same `WSTRNCMP(cmd, "scp", 3)` and is the only thing that decides whether the SCP path runs. `wolfSSH_accept()` diverts to `WS_SCP_INIT` on that test in `src/ssh.c`, and wolfsshd serves SCP only from that return, so "scpX" is already an SCP command as far as the library is concerned. The callback has to accept exactly what the divert will hand off.

Tightening it to "scp" or "scp<whitespace>" would make the callback the stricter of the two: on a build with `WOLFSSH_SCP` and without `WOLFSSH_SHELL`, the daemon would answer CHANNEL_FAILURE to an exec that the divert would otherwise have passed to `SCP_Subsystem()`. That changes what the daemon serves, which this change is meant to leave alone. If the prefix is too loose, the place to tighten it is `ChannelCommandIsScp()`, with the callback kept in step.

The read is bounded either way: `GetStringAlloc()` NUL-terminates the command, so a command shorter than three bytes stops the comparison at the terminator rather than reading past it.

File: src/ssh.c [hunk 2]
```diff
@@ -4931,6 +4913,32 @@ int wolfSSH_CTX_SetChannelReqSubsysCb(WOLFSSH_CTX* ctx,
...
+{
+    int ret = WS_SSH_NULL_E;
+
+    if (ssh != NULL) {
+        ssh->appChannels = (enable != 0);
+        ret = WS_SUCCESS;
+    }
+
+    return ret;
+}
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply src/ssh.c:5856]: Addressed on master, and this hunk is no longer part of this PR after the rebase. The `wolfssh/ssh.h` comment no longer claims late enable has no effect; it now says turning it on later still applies to the channel requests that follow, and only that it cannot move where `wolfSSH_accept()` returns on a session already past the user-auth stop. `TestAppChannelsLateEnableReturns()` in `tests/regress.c` covers that.

File: wolfssh/agent.h [hunk 3]
```diff
@@ -181,6 +181,14 @@ WOLFSSH_API int wolfSSH_CTX_set_agent_cb(WOLFSSH_CTX* ctx,
 WOLFSSH_API int wolfSSH_set_agent_cb_ctx(WOLFSSH* ssh, void* ctx);
 WOLFSSH_API int wolfSSH_CTX_AGENT_enable(WOLFSSH_CTX* ctx, byte isEnabled);
 WOLFSSH_API int wolfSSH_AGENT_enable(WOLFSSH* ssh, byte isEnabled);
+/* Server side. Opens the auth-agent@openssh.com channel back to the client
+ * once an auth-agent-req@openssh.com request has set the session up for it.
+ * wolfSSH_accept() does this itself on the default path; an application that
+ * drives its own channels returns from accept() before that point and calls
+ * this instead. Idempotent, so it is safe to poll while waiting for the
+ * peer's request. Returns WS_SUCCESS, or WS_BAD_ARGUMENT when the session
+ * never asked for agent forwarding. */
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply wolfssh/agent.h]: Addressed on master, and `wolfssh/agent.h` is no longer part of this PR after the rebase. The comment on `wolfSSH_AGENT_ChannelOpen()` now enumerates the full set: `WS_SUCCESS`, `WS_BAD_ARGUMENT`, `WS_WANT_READ`/`WS_WANT_WRITE`, `WS_FATAL_ERROR` with `WS_DISCONNECT` in `ssh->error`, `WS_SSH_NULL_E`, `WS_MEMORY_E`, or whatever the send reports.

File: wolfssh/ssh.h [hunk 4]
```diff
@@ -452,6 +452,29 @@ WOLFSSH_API int wolfSSH_CTX_SetChannelReqSubsysCb(WOLFSSH_CTX* ctx,
...
+ * or subsystem request with no callback registered is then rejected: with
+ * accept() already returned, nothing is left to service it.
+ *
+ * Set it on the context before wolfSSH_new(), or on a session before the
+ * first wolfSSH_accept() call. Turning it on once accept() has established
+ * the session has no effect on that session.
+ *
+ * The mode drives the session channels itself, so it does not combine with
+ * the built-in wolfSSH_SFTP_accept() and WS_SCP_INIT entry points; an
+ * application using those leaves this off. */
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply wolfssh/ssh.h]: Addressed on master, and `wolfssh/ssh.h` is no longer part of this PR after the rebase. The comment now states the combinations as they are: `WS_SCP_INIT` is off the table because accept() never reaches the built-in SCP entry point in this mode, while `wolfSSH_SFTP_accept()` still serves a session channel the subsystem callback granted sftp on, and returns `WS_INVALID_STATE_E` without recording an error when called ahead of that.

File: apps/wolfsshd/wolfsshd.c [hunk 5]
```diff
@@ -471,6 +471,92 @@ static void CleanupCTX(WOLFSSHD_CONFIG* conf, WOLFSSH_CTX** ctx,
...
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply apps/wolfsshd/wolfsshd.c]: Agreed, and fixed. `SessionRequestCb()` now requires `scp` to stand as its own token -- the command is either exactly `scp`, or `scp` followed by a space -- tested against the wire length rather than as a prefix, so `scpbackup` is refused and a command with an embedded NUL is treated as some other command.

Worth noting the gate only decides policy here: `ChannelCommandIsScp()` in `src/wolfscp.c` still classifies by the same 3-byte prefix, which is Fenrir finding 11664. The divert requires `sessionGranted`, so a refusal at this gate wins regardless of how the classifier would answer.

File: apps/wolfsshd/wolfsshd.c [hunk 6]
```diff
@@ -471,6 +471,101 @@ static void CleanupCTX(WOLFSSHD_CONFIG* conf, WOLFSSH_CTX** ctx,
...
+                reason = "exec request carried no command";
+                break;
+            }
+        #ifdef WOLFSSH_SCP
+            {
+                word32 cmdSz = wolfSSH_ChannelGetSessionCommandSz(channel);
+
+                /* "scp" must stand as its own token; a prefix match
+                 * grants "scpbackup", and a NUL makes it another. */
+                if (cmdSz >= (word32)WSTRLEN("scp")
```
Comment (padelsbach) [inline apps/wolfsshd/wolfsshd.c]: Consider doing `const  size_t scp_strlen = WSTRLEN("scp");` prior to the `if` for unification and to remove the literal `3`

Alternatively, why not use `ChannelCommandIsScp` (or a public wrapper) here?

File: apps/wolfsshd/wolfsshd.c [hunk 7]
```diff
@@ -471,6 +471,101 @@ static void CleanupCTX(WOLFSSHD_CONFIG* conf, WOLFSSH_CTX** ctx,
...
+        case WOLFSSH_SESSION_EXEC:
+            if (cmd == NULL) {
+                reason = "exec request carried no command";
+                break;
+            }
+        #ifdef WOLFSSH_SCP
+            {
+                word32 cmdSz = wolfSSH_ChannelGetSessionCommandSz(channel);
+
+                /* "scp" must stand as its own token; a prefix match
```
Comment (padelsbach) [inline apps/wolfsshd/wolfsshd.c]: `ChannelCommandIsScp` seems to match `scpbackup` in `wolfSSH_accept`. Is this mismatch ok?

File: apps/wolfsshd/wolfsshd.c [hunk 8]
```diff
@@ -471,6 +471,101 @@ static void CleanupCTX(WOLFSSHD_CONFIG* conf, WOLFSSH_CTX** ctx,
...
+        case WOLFSSH_SESSION_TERMINAL:
+        default:
+            reason = "unsupported session type";
+            break;
+    }
+
+    /* One program start per channel, as RFC 4254 section 6.5 allows. This
+     * request's grant is recorded once the callback returns, so a flag
+     * already set is an earlier request's. */
+    if (!rej && channel->sessionGranted) {
```
Comment (padelsbach) [inline apps/wolfsshd/wolfsshd.c]: Do we normally allow/recommend app code to use data from `WOLFSSH_CHANNEL`? Quick search of the `apps/` doesn't turn up anything. Maybe add a getter?

**Review by padelsbach — COMMENTED** [review]
Couple minor points, but looks good overall

File: apps/wolfsshd/wolfsshd.c (hunk 8 above)
Comment (ejohnstown) [reply apps/wolfsshd/wolfsshd.c]: Added `wolfSSH_ChannelGetSessionGranted()` next to the other channel getters in ssh.h, and the callback uses it.

File: apps/wolfsshd/wolfsshd.c (hunk 7 above)
Comment (ejohnstown) [reply apps/wolfsshd/wolfsshd.c]: No, and thanks, it's a real one. An exec of scpbackup foo was granted by the callback as an ordinary exec, then `wolfSSH_accept()` prefix-matched it, returned `WS_SCP_INIT`, and wolfsshd ran `SCP_Subsystem()` on it. Same for `scp\0junk`, where  `ChannelCommandIsScp()` used `WSTRLEN` and stopped at the NUL.

Fixed by having one definition instead of two: `wolfSSH_ChannelCommandIsScp()` applies the token rule against the recorded command size, and both the divert in `wolfSSH_accept()` and the callback call it, so they can't drift apart again.

File: apps/wolfsshd/wolfsshd.c (hunk 6 above)
Comment (ejohnstown) [reply apps/wolfsshd/wolfsshd.c]: Since wolfSSHd is an application, switched to `wolfSSH_ChannelCommandIsScp()`.

File: src/wolfscp.c [hunk 9]
```diff
@@ -1040,20 +1040,25 @@ WOLFSSH_API int wolfSSH_SetScpErrorMsg(WOLFSSH* ssh, const char* message)
...
         return WS_BAD_ARGUMENT;
 
-    cmd = wolfSSH_GetSessionCommand(ssh);
-    if (cmd != NULL && WSTRLEN(cmd) >= 3) {
-        if (WSTRNCMP(cmd, "scp", 3) == 0)
-            ret = 1;
+    cmd = wolfSSH_ChannelGetSessionCommand(channel);
+    cmdSz = wolfSSH_ChannelGetSessionCommandSz(channel);
+
+    if (cmd != NULL && cmdSz >= scpSz
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply src/wolfscp.c:1059]: Agreed on the mechanism, and fixed. `wolfSSH_ChannelCommandIsScp()` now refuses a command carrying a NUL anywhere in the recorded command size, so `scp -t /d\0x` is an ordinary exec and wolfsshd answers CHANNEL_FAILURE rather than serving the truncated `scp -t /d`. The token table in `tests/regress.c` gains the NUL-past-the-boundary case and a NUL counted at the end, and the contract in `wolfssh/wolfscp.h` now states the rule the code enforces -- it had claimed that testing against the recorded size kept an embedded NUL from hiding the rest of the command, which held only for a NUL at the boundary.

On severity, nothing here crosses a privilege boundary. Every consumer truncates at the same NUL -- this gate, the divert in `wolfSSH_accept()`, `ParseScpCommand()`, and the wolfsshd log -- so the grant and the service agree with each other, and the hidden bytes are discarded rather than parsed. A forced command is substituted for the client's command rather than compared against it, so it is not a policy bypass either: an attacker gains nothing over sending `scp -t /d` directly. Worth fixing because the header promised the property and because the sftp gate in the same callback already matches on length, not because it is reachable as an escalation.

**Review by padelsbach — APPROVED** [review]
